shared/actions/signup.js

```diff
@@ -1,366 +1,304 @@
-        passphraseError = new HiddenString('Passphrases must match')
-      } else if (passphrase1.length < 6) {
-        passphraseError = new HiddenString('Passphrase must be at least 6 characters long')
-      }
-
-      if (passphraseError) {
-        dispatch(
-          SignupGen.createCheckPassphraseError({
-            passphraseError,
-          })
-        )
-      } else {
-        dispatch(
-          SignupGen.createCheckPassphrase({
-            passphrase: new HiddenString(passphrase1),
-          })
-        )
-        dispatch(nextPhase())
-      }
-
-      resolve()
+const submitDevicenameSuccess = () => Saga.put(SignupGen.createSignup())
+const submitDevicenameError = (err: RPCError, action: SignupGen.SubmitDevicenamePayload) => {
+  logger.warn('device name is invalid: ', err)
+  return Saga.put(
+    SignupGen.createSubmitDevicenameDoneError({
+      devicename: action.payload.devicename,
+      error: `Device name is invalid: ${err.desc}.`,
     })
-    return p
-  }
+  )
 }
 
-function submitDeviceName(deviceName: string, skipMail?: boolean, onDisplayPaperKey?: () => void) {
-  return (dispatch: Dispatch) => {
-    const p = new Promise((resolve, reject) => {
-      // TODO do some checking on the device name - ideally this is done on the service side
-      let deviceNameError = null
-      if (trim(deviceName).length === 0) {
-        deviceNameError = 'Device name must not be empty.'
-      }
-
-      if (deviceNameError) {
-        dispatch(
-          SignupGen.createSubmitDeviceNameError({
-            deviceNameError: deviceNameError || '',
-          })
-        )
-      } else {
-        RPCTypes.deviceCheckDeviceNameFormatRpcPromise({
-          name: deviceName,
-          waitingHandler: isWaiting => {
-            dispatch(SignupGen.createWaiting({waiting: isWaiting}))
-          },
-        })
-          .then(() => {
-            if (deviceName) {
-              dispatch(SignupGen.createSubmitDeviceName({deviceName}))
+const signup = (action: SignupGen.SignupPayload, state: TypedState) => {
+  const {email, username, inviteCode, passphrase, devicename} = state.signup
 
-              const signupPromise = dispatch(signup(skipMail || false, onDisplayPaperKey))
-              if (signupPromise) {
-                signupPromise.then(resolve, reject)
-              } else {
-                throw new Error('did not get promise from signup')
-              }
-            }
-          })
-          .catch(err => {
-            logger.warn('device name is invalid: ', err)
-            dispatch(
-              SignupGen.createSubmitDeviceNameError({
-                deviceNameError: `Device name is invalid: ${err.desc}.`,
-              })
-            )
-            resolve()
-          })
-      }
-    })
-    return p
+  if (!email || !username || !inviteCode || !passphrase || !passphrase.stringValue() || !devicename) {
+    throw new Error('Missing data for signup')
```

Comment:
Probably worthy of a `logger.error`. I think logs at all the error cases in this file would be helpful in diagnosing stuff (esp. user confusion)